### PR TITLE
DBZ-1491 Extract default values from schema

### PR DESCRIPTION
**Pull request by grzegorz8**
The goal is to enrich schema metadata with column default value.

### CODE REVIEW

File: debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerConnection.java [hunk 1]
```diff
@@ -62,6 +68,59 @@
...
+            String datetime = rawValue.substring(0, rawValue.length() - 6);
+            String offset = rawValue.substring(rawValue.length() - 6);
+            int offsetSignum = offset.charAt(0) == '-' ? -1 : 1;
+            int offsetHours = Integer.parseInt(offset.substring(1, 3));
+            int offsetMinutes = Integer.parseInt(offset.substring(4, 6));
+            int minutesOffset = offsetSignum * (offsetHours * 60 + offsetMinutes);
+            return DateTimeOffset.valueOf(Timestamp.valueOf(datetime), minutesOffset);
+        });
+        DEFAULT_VALUE_MAPPERS.put("smalldatetime", v -> Timestamp.valueOf(v.substring(2, v.length() - 2))); // Sample value: ('2019-01-01 00:00:00')
+        DEFAULT_VALUE_MAPPERS.put("time", v -> new Timestamp(Time.valueOf(v.substring(2, v.length() - 2)).toLocalTime().toSecondOfDay() * 1000L));  // Sample value: ('00:00:56')
```
Comment (grzegorz8) [inline debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerConnection.java]: https://docs.microsoft.com/en-us/sql/t-sql/data-types/time-transact-sql?view=sql-server-ver15

Nonoseconds are not covered.

**Comment by jpechane** [comment]
@grzegorz8 Good job! I've few comments still

- Could you please add a test for `ALTER TABLE` tha modifies the default value?
- Are not default values for date/time depending on local settings of the database?
- Could you please provide and to end tests? Thus verifing that the default value in connect schema is set correctly and verify what happens if you provide null value for tables with NULL/NOT NULL with or without DEFAULT value set?

File: debezium-core/src/main/java/io/debezium/jdbc/JdbcConnection.java [hunk 2]
```diff
@@ -1074,13 +1090,37 @@ public void readSchema(Tables tables, String databaseCatalog, String schemaNameP
...
+        if (mapper == null) {
+            LOGGER.warn("Mapper for type '{}' not found.", dataType);
+            return Optional.empty();
+        }
+
+        try {
+            return Optional.of(mapper.parse(defaultValue));
+        }
+        catch (Exception e) {
+            LOGGER.error("Cannot parse column default value '{}' to type '{}'.", defaultValue, dataType, e);
```
Comment (jpechane) [inline debezium-core/src/main/java/io/debezium/jdbc/JdbcConnection.java]: This should probably be only `WARN`as to connector continue to run

**Comment by grzegorz8** [comment]
Sure, I'll apply improvements as soon as possible.
>     * Are not default values for date/time depending on local settings of the database?
Any ideas how to parse the value properly in this case?

**Comment by jpechane** [comment]
@grzegorz8 Does SQL Server support query that does typecasting? If yes is it possible send query like
`SELECT TODATE('<defaultvalue'>)`? This should handle locale settings, should not it?

**Comment by grzegorz8** [comment]
Even better, there is PARSE() function: https://docs.microsoft.com/en-us/sql/t-sql/functions/parse-transact-sql?view=sql-server-ver15

**Comment by gunnarmorling** [comment]
@jpechane, @grzegorz8, where aer we with this one? One general question: is this prone to race conditions in case of default value changes / out-of-bands reads? Something worth documenting?

**Comment by jpechane** [comment]
@gunnarmorling The matter of loclized temporal values is still open. Regarding the race condiiton - ye there is possibility of it but IMHO if you follow the required database schem change procedure then the issue should not be there.

**Comment by grzegorz8** [comment]
I am writing end-to-end tests now. Thanks to them I discovered that temporal values are converted to long values instead of Timestamp in ConnectSchema. As a result, I need some refactoring in previous changes too :)

Sorry it takes so long, I am able to develop it only during weekends.

**Comment by jpechane** [comment]
@grzegorz8 Please don't feel pressed. We appreciate all contributions and I understand your constraints, We are just planning the next release so we are checking what will or will not be the part of it. Take as much time as you need!

**Comment by gunnarmorling** [comment]
Hey @grzegorz8, just checking in on this one. Any news? Thanks!

**Comment by grzegorz8** [comment]
It's almost ready, however, there are still a few tests failing, which I need to investigate.

**Comment by gunnarmorling** [comment]
Hey @grzegorz8, it has been a while, howdy? Are you still planning to continue with this one? Or should someone else take over? Thanks for a quick update.

**Comment by grzegorz8** [comment]
@gunnarmorling I pushed my changes but I forgot to write it's ready to review.
I see one test failed in debezium-connector-postgres and "SqlServerChangeTableSetIT.addDefaultValue:670 expecting actual value not to be null" but I am not able to reproduce it locally.

**Comment by gunnarmorling** [comment]
> it's ready to review.

Ah, cool. Can you rebase and force-push to this PR? This will trigger a new
build; we addressed a few intermittent test failures lately.

**Comment by jpechane** [comment]
Rebased to the latest master, closing in favour of https://github.com/debezium/debezium/pull/1564

File: debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerDefaultValueConverter.java [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+            return valueConverter.convert(defaultValue);
+        }
+        return defaultValue;
+    }
+
+    private Map<String, DefaultValueMapper> createDefaultValueMappers() {
+        final Map<String, DefaultValueMapper> result = new HashMap<>();
+
+        // Exact numbers
+        result.put("bigint", v -> Long.parseLong(v.substring(2, v.length() - 3))); // Sample value: ((3147483648.))
```
Comment (mrjrieke) [inline debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerDefaultValueConverter.java:117]: Not safe.

File: debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerDefaultValueConverter.java [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+        result.put("int", v -> Integer.parseInt(v.substring(2, v.length() - 2))); // Sample value: ((2147483647))
+        result.put("smallint", v -> Short.parseShort(v.substring(2, v.length() - 2))); // Sample value: ((32767))
+        result.put("tinyint", v -> Short.parseShort(v.substring(2, v.length() - 2))); // Sample value: ((255))
+        result.put("bit", v -> v.equals("((1))")); // Either ((1)) or ((0))
+        result.put("decimal", v -> new BigDecimal(v.substring(2, v.length() - 2))); // Sample value: ((100.12345))
```
Comment (mrjrieke) [inline debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerDefaultValueConverter.java:122]: not safe.

**Pull request closed without merging**